### ISSUE TITLE
[Fastlane/Apple 2FA] Session loaded from environment variable is not valid. 

### ISSUE THREAD

**Issue by BenjaminSAILcloud** [microsoft/app-store-vsts-extension#180]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: Apple App Store Release  

## Environment
- Server - Azure Pipelines or TFS on-premises?
    
    - If using TFS on-premises, provide the version: 
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: 
private project

- Agent - Hosted or Private: 
    
    - If using Hosted agent, provide agent queue name:

    - If using private agent, provide the OS of the machine running the agent and the agent version: 

## Issue Description

The 2FA app-specific password does not work.  
I used this pipeline and it worked only once. After that it stopped working. I tried creating another app-specific password from my apple account and set the task accordingly but I didn't work either.

### Task logs
<details>
  <summary>Logs</summary>

```shell
    ==============================================================================
    Task         : Apple App Store Release
    Description  : Release an app to TestFlight or the Apple App Store
    Version      : 1.176.0
    Author       : Microsoft Corporation
    Help         : [More Information](https://marketplace.visualstudio.com/items?itemName=ms-vsclient.app-store)
    ==============================================================================
    /usr/local/opt/ruby/bin/gem install fastlane
    Successfully installed fastlane-2.162.0
    Parsing documentation for fastlane-2.162.0
    Installing ri documentation for fastlane-2.162.0
    Done installing documentation for fastlane after 22 seconds
    1 gem installed

    /usr/local/opt/ruby/bin/gem update fastlane -i /Users/runner/.gem-cache
    Updating installed gems
    Nothing to update

    /usr/local/lib/ruby/gems/2.6.0/bin/fastlane pilot upload -u email@gmail.com -i /Users/runner/work/1/s/output/iphoneos/Release/SAILcloudBox.ipa --changelog # 2.0.0.61

    update/npm packages
     -q 123456789 -r xxxxxxxxxxxx -a xxx.xxxxxxxx.xxx --skip_submission true
    [09:54:08]: Seems like launching fastlane takes a while - please run
    [09:54:08]: 
    [09:54:08]: $ [sudo] gem cleanup
    [09:54:08]: 
    [09:54:08]: to uninstall outdated gems and make fastlane launch faster
    [09:54:08]: Alternatively it's recommended to start using a Gemfile to lock your dependencies
    [09:54:08]: To get started with a Gemfile, run
    [09:54:08]: 
    [09:54:08]: $ bundle init
    [09:54:08]: $ echo 'gem "fastlane"' >> Gemfile
    [09:54:08]: $ bundle install
    [09:54:08]: 
    [09:54:08]: After creating the Gemfile and Gemfile.lock, commit those files into version control
    [09:54:08]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
    [09:54:16]: Sending anonymous analytics information
    [09:54:16]: Learn more at https://docs.fastlane.tools/#metrics
    [09:54:16]: No personal or sensitive data is sent.
    [09:54:16]: You can disable this by adding `opt_out_usage` at the top of your Fastfile
    [09:54:16]: Login to App Store Connect (email@gmail.com)
    Session loaded from environment variable is not valid. Continuing with normal login.
    Two-factor Authentication (6 digits code) is enabled for account 'email@gmail.com'
    More information about Two-factor Authentication: https://support.apple.com/en-us/HT204915

    If you're running this in a non-interactive session (e.g. server or CI)
    check out https://github.com/fastlane/fastlane/tree/master/spaceship#2-step-verification

    Please select a trusted phone number to send code to:
    1. +33 •• •• •• •• 93
    2. +33 •• •• •• •• 95
    ?  
    ##[error]The operation was canceled.
    Finishing: Publish to the App Store TestFlight track

```
</details>

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

### Error logs

```
    Session loaded from environment variable is not valid. Continuing with normal login.
    Two-factor Authentication (6 digits code) is enabled for account 'email@gmail.com'
    More information about Two-factor Authentication: https://support.apple.com/en-us/HT204915

    If you're running this in a non-interactive session (e.g. server or CI)
    check out https://github.com/fastlane/fastlane/tree/master/spaceship#2-step-verification

    Please select a trusted phone number to send code to:
    1. +33 •• •• •• •• XX
    2. +33 •• •• •• •• XX
    ?  
    ##[error]The operation was canceled.
    Finishing: Publish to the App Store TestFlight track

```

**Comment by anatolybolshakov** [comment]
Hi @BenjaminSAILcloud to disable 2FA using app specific password - could you please check following conditions?
- app specific apple id should be provided (-p "your apple id" as fastlaneArguments input vaue)
- shouldSkipWaitingForProcessing: true
- isTwoFactorAuth: true
- there should not be any changelog

**Comment by BenjaminSAILcloud** [comment]
@anatolybolshakov I tried adding the -p arg with the apple ID, but it does not work either. I get this error : 
`[!] 'apple_id' value is incorrect. The correct value should be taken from Apple ID property in the App Information section in App Store Connect.`

But I don't know why in the command line, the task replaces the apple ID with `***`, is this a security feature ?
`Error setting value '***' for option 'apple_id'`
with command : 
`/usr/local/lib/ruby/gems/2.6.0/bin/fastlane pilot upload -u *** -i /Users/runner/work/1/s/output/iphoneos/Release/app.ipa -q 123456789 -r Team -a id.company.com --skip_submission true --skip_waiting_for_build_processing true -p ***`

Also where do you add the attribute `isTwoFactorAuth` ? It does not appear in the doc...

<details>
  <summary>❯ fastlane pilot upload --help</summary>

```shell
    Usage: fastlane pilot upload

  Uploads a new binary to Apple TestFlight

  Options:
    --api_key_path STRING Path to your App Store Connect API Key JSON file
(https://docs.fastlane.tools/app-store-connect-api/#using-fastlane-api-key-json-file) (PILOT_API_KEY_PATH)
    --api_key STRING     Your App Store Connect API Key information
(https://docs.fastlane.tools/app-store-connect-api/#use-return-value-and-pass-in-as-an-option) (PILOT_API_KEY)
    -u, --username STRING Your Apple ID Username (PILOT_USERNAME)
    -a, --app_identifier STRING The bundle identifier of the app to upload or manage testers (optional) (PILOT_APP_IDENTIFIER)
    -m, --app_platform STRING The platform to use (optional) (PILOT_PLATFORM)
    -p, --apple_id STRING Apple ID property in the App Information section in App Store Connect (PILOT_APPLE_ID)
    -i, --ipa STRING     Path to the ipa file to upload (PILOT_IPA)
    --demo_account_required [VALUE] Do you need a demo account when Apple does review? (DEMO_ACCOUNT_REQUIRED)
    --beta_app_review_info STRING Beta app review information for contact info and demo account (PILOT_BETA_APP_REVIEW_INFO)
    --localized_app_info STRING Localized beta app test info for description, feedback email, marketing url, and privacy policy
(PILOT_LOCALIZED_APP_INFO)
    -d, --beta_app_description STRING Provide the 'Beta App Description' when uploading a new build (PILOT_BETA_APP_DESCRIPTION)
    -n, --beta_app_feedback_email STRING Provide the beta app email when uploading a new build (PILOT_BETA_APP_FEEDBACK)
    --localized_build_info STRING Localized beta app test info for what's new (PILOT_LOCALIZED_BUILD_INFO)
    -w, --changelog STRING Provide the 'What to Test' text when uploading a new build (PILOT_CHANGELOG)
    -s, --skip_submission [VALUE] Skip the distributing action of pilot and only upload the ipa file (PILOT_SKIP_SUBMISSION)
    -z, --skip_waiting_for_build_processing [VALUE] If set to true, the `distribute_external` option won't work and no build will be distributed to
testers. (You might want to use this option if you are using this action on CI and have to pay for 'minutes used' on your CI plan). If set to `true` and
a changelog is provided, it will partially wait for the build to appear on AppStore Connect so the changelog can be set, and skip the remaining
processing steps (PILOT_SKIP_WAITING_FOR_BUILD_PROCESSING)
    -x, --update_build_info_on_upload [VALUE] **DEPRECATED!** Update build info immediately after validation. This is deprecated and will be removed in
a future release. App Store Connect no longer supports setting build info until after build processing has completed, which is when build info is
updated by default (PILOT_UPDATE_BUILD_INFO_ON_UPLOAD)
    -D, --distribute_only [VALUE] Distribute a previously uploaded build (equivalent to the `fastlane pilot distribute` command) (PILOT_DISTRIBUTE_ONLY)

    -X, --uses_non_exempt_encryption [VALUE] Provide the 'Uses Non-Exempt Encryption' for export compliance. This is used if there is
'ITSAppUsesNonExemptEncryption' is not set in the Info.plist (PILOT_USES_NON_EXEMPT_ENCRYPTION)
    --distribute_external [VALUE] Should the build be distributed to external testers? (PILOT_DISTRIBUTE_EXTERNAL)
    --notify_external_testers [VALUE] Should notify external testers? (PILOT_NOTIFY_EXTERNAL_TESTERS)
    --app_version STRING The version number of the application build to distribute. If the version number is not specified, then the most recent build
uploaded to TestFlight will be distributed. If specified, the most recent build for the version number will be distributed (PILOT_APP_VERSION)
    --build_number STRING The build number of the application build to distribute. If the build number is not specified, the most recent build is
distributed (PILOT_BUILD_NUMBER)
    --expire_previous_builds [VALUE] Should expire previous builds? (PILOT_EXPIRE_PREVIOUS_BUILDS)
    -f, --first_name STRING The tester's first name (PILOT_TESTER_FIRST_NAME)
    -l, --last_name STRING The tester's last name (PILOT_TESTER_LAST_NAME)
    -e, --email STRING   The tester's email (PILOT_TESTER_EMAIL)
    -c, --testers_file_path STRING Path to a CSV file of testers (PILOT_TESTERS_FILE)
    -g, --groups ARRAY   Associate tester to one group or more by group name / group id. E.g. `-g "Team 1","Team 2"` (PILOT_GROUPS)
    -q, --team_id [VALUE] The ID of your App Store Connect team if you're in multiple teams (PILOT_TEAM_ID)
    -r, --team_name STRING The name of your App Store Connect team if you're in multiple teams (PILOT_TEAM_NAME)
    --dev_portal_team_id STRING The short ID of your team in the developer portal, if you're in multiple teams. Different from your iTC team ID!
(PILOT_DEV_PORTAL_TEAM_ID)
    --itc_provider STRING The provider short name to be used with the iTMSTransporter to identify your team. This value will override the automatically
detected provider short name. To get provider short name run `pathToXcode.app/Contents/Applications/Application\
Loader.app/Contents/itms/bin/iTMSTransporter -m provider -u 'USERNAME' -p 'PASSWORD' -account_type itunes_connect -v off`. The short names of providers
should be listed in the second column (PILOT_ITC_PROVIDER)
    -k, --wait_processing_interval INTEGER Interval in seconds to wait for App Store Connect processing (PILOT_WAIT_PROCESSING_INTERVAL)
    --wait_for_uploaded_build [VALUE] **DEPRECATED!** No longer needed with the transition over to the App Store Connect API - Use version info from
uploaded ipa file to determine what build to use for distribution. If set to false, latest processing or any latest build will be used
(PILOT_WAIT_FOR_UPLOADED_BUILD)
    -b, --reject_build_waiting_for_review [VALUE] Expire previous if it's 'waiting for review' (PILOT_REJECT_PREVIOUS_BUILD)
```
</details>

**Comment by anatolybolshakov** [comment]
@BenjaminSAILcloud did you pass apple id from Apple ID property in the App Information section in App Store Connect? This has numeric format and differs from user Apple ID.
Yep, hiding of sensitive data is expected - it doesn't affect passed values and only related to displaying of sensitive params in logs
isTwoFactorAuth is the pipeline task input - could you please check that it's being set in yaml file?

**Comment by BenjaminSAILcloud** [comment]
Ok great, the problem was indeed the Apple ID.
This is working now, thanks a lot for the help.
Closing this.
